### PR TITLE
feat: direct swarming

### PR DESCRIPTION
**Pull request by nduchak**

### CODE REVIEW

File: src/definitions.ts [hunk 1]
```diff
@@ -260,7 +260,7 @@ export interface AgreementSizeExceededPayload {
 
 export interface MultiaddrAnnouncementPayload {
   agreementReference: string
-  multiaddr: string
+  peerId: string
```
Comment (nduchak) [inline src/definitions.ts:263]: Can we change payload here from multiadress to peerId?

**Comment by nduchak** [comment]
Closes #64

File: src/definitions.ts (hunk 1 above)
Comment (AuHau) [reply src/definitions.ts:263]: Sure, sounds good!

File: src/utils.ts [hunk 2]
```diff
@@ -312,3 +313,8 @@ export default abstract class BaseCommand extends Command {
     }
   }
 }
+
+export async function getPeerIdByAgreement (agreementReference: string): Promise<string | undefined> {
```
Comment (AuHau) [inline src/utils.ts:317]: Hmm, I think we should also prune the used PeerIds as I think they are one-time only usable information. Hence maybe after returning the PeerId here, I would delete the record from the DB? What do you think?

**Review by AuHau — COMMENTED** [review]
Almost LGTM! One suggestion, but then I think it is ready to be merged.

File: src/utils.ts (hunk 2 above)
Comment (nduchak) [reply src/utils.ts:317]: Yea, good idea

**Review by vojtechsimetka — COMMENTED** [review]
Fake `DirectAddressModel` records are there forever, right? So this is potential attack filling in the DB of a pinner with fake requests.

**Comment by nduchak** [comment]
> Fake `DirectAddressModel` records are there forever, right? So this is potential attack filling in the DB of a pinner with fake requests.
It's removed right after we retrieve it from db: https://github.com/rsksmart/rif-storage-pinner/pull/185/files#diff-e1495d267619047a7cca5cfe8f692729R319

**Comment by AuHau** [comment]
Hmm but Vojtech is talking about the records that won't get used for any agreements. Yeah, we should add a timestamp there (just use the Sequelize mechanism, eq: `timestamps: true`) and maybe as part of GC routine delete older records than configured timespan.

**Comment by nduchak** [comment]
I see, thanks
Will add it soon

File: src/migrations/scripts/1597928323618-initial.ts [hunk 3]
```diff
@@ -83,6 +83,44 @@ function migrationCommands (transaction: any): Commands {
         }
       ]
     },
+    {
+      fn: 'createTable',
```
Comment (AuHau) [inline src/migrations/scripts/1597928323618-initial.ts:87]: Uff I would say this should be a new migrations. The idea of migrations are that they are immutable and don't change, but only add-on. Yes it was not released yet, actually it was already as the pre-releases, so yes this should be a new migration...

File: src/processor/marketplace-events/index.ts [hunk 4]
```diff
@@ -56,7 +56,10 @@ export class MarketplaceEventsProcessor extends EventProcessor {
       )
 
       this.manager = manager
-      this.gcHandler = errorHandler(collectPinsClosure(this.manager), loggingFactory('gc'))
+      this.gcHandler = composeGc([
```
Comment (AuHau) [inline src/processor/marketplace-events/index.ts:59]: Hmm I would refactor this to have it as part of the `gc.ts` file and here only calling the main function. If you will be searching what is GCed, you won't be searching for it here... Also then you don't have to compose it twice, here and in the blockchain events processor...

**Review by AuHau — CHANGES_REQUESTED** [review]
Two things to address 😇 But almost there!

File: src/migrations/scripts/1597928323618-initial.ts (hunk 3 above)
Comment (nduchak) [reply src/migrations/scripts/1597928323618-initial.ts:87]: Sure

File: src/processor/marketplace-events/index.ts (hunk 4 above)
Comment (nduchak) [reply src/processor/marketplace-events/index.ts:59]: Agree. Good point

File: src/@types/parse-duration/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,4 @@
+declare module 'parse-duration' {
```
Comment (AuHau) [inline src/@types/parse-duration/index.d.ts:1]: Why are you adding type declarations when it has typings? I don't think they are needed. I have already installed the package for my branch as well and everything works alright...

**Review by AuHau — APPROVED** [review]
Please fix the CI and remove the unnecessary typings and then we are ready to merge ;-)

File: src/@types/parse-duration/index.d.ts (hunk 5 above)
Comment (nduchak) [reply src/@types/parse-duration/index.d.ts:1]: Hmm..
WIll double check

**Pull request merged**